### ISSUE TITLE
Tacotron2 fine-tune problem

### ISSUE THREAD

**Issue by unparalleled-ysj** [TensorSpeech/TensorFlowTTS#194]
Hello, I trained a tacotron2 model with the A dataset. Based on this, I want to use the B dataset for fine-tuning, but something wrong:
![1](https://user-images.githubusercontent.com/39238619/89749643-32abfc80-dafb-11ea-8a37-42e8eaa25d12.png)
![1_alignment](https://user-images.githubusercontent.com/39238619/89749675-47889000-dafb-11ea-8f55-b6420045b0a5.png)

this is tensorboard:
<img width="1072" alt="截屏2020-08-10 上午9 35 44" src="https://user-images.githubusercontent.com/39238619/89749666-42c3dc00-dafb-11ea-91f2-47507a09ee1a.png">

It seems that the data does not flow into the model correctly? But there is no error message, and the data is normally read and loaded into the memory

My fine-tuned command is：
CUDA_VISIBLE_DEVICES=0 python examples/tacotron2/train_tacotron2.py --train-dir training_data/B/train/ --dev-dir training_data/B/valid/ --outdir examples/tacotron2/exp/B/ --config examples/tacotron2/conf/tacotron2.v1.yaml --use-norm 1 --mixed_precision 0 --resume examples/tacotron2/exp/B/checkpoints/ckpt-200000
where examples/tacotron2/exp/B/checkpoints/ckpt-200000 is copied from examples/tacotron2/exp/A/checkpoints/

But when I use the B dataset to train from scratch, every is fine:
CUDA_VISIBLE_DEVICES=0 python examples/tacotron2/train_tacotron2.py --train-dir training_data/B/train/ --dev-dir training_data/B/valid/ --outdir examples/tacotron2/exp/B/ --config examples/tacotron2/conf/tacotron2.v1.yaml --use-norm 1 --mixed_precision 0 --resume ""

![new](https://user-images.githubusercontent.com/39238619/89750096-1ad57800-dafd-11ea-8387-f6bcfac06187.png)
![new_alignment](https://user-images.githubusercontent.com/39238619/89750110-2a54c100-dafd-11ea-8a89-17fae95a66e1.png)

**Comment by dathudeptrai** [comment]
@unparalleled-ysj let me expained the problem:

Fine-tune and resume are different. To fine-tune you just need add `tacotron2.load_weight(PATH_WEIGHT_MODEL_A)` after this line https://github.com/TensorSpeech/TensorFlowTTS/blob/master/examples/tacotron2/train_tacotron2.py#L425. The training progress sill begin at epoch = 0, iterations = 0. Use pretrained from model A may let the model convergence faster. (but not sure, training from scratch still better in some cases)

**Comment by ZDisket** [comment]
@dathudeptrai Why not include a `--pretrained` or `--finetune` command line option in the scripts that loads weights?

**Comment by dathudeptrai** [comment]
@ZDisket i will do :(, just forgot about it :(

**Comment by unparalleled-ysj** [comment]
@dathudeptrai  get it , thank you for your reply

**Comment by unparalleled-ysj** [comment]
@dathudeptrai I followed the method you said to fine-tune, but this problem still occurs。
this is my code
`with STRATEGY.scope():
        # define model.
        tacotron_config = Tacotron2Config(**config["tacotron2_params"])
        tacotron2 = TFTacotron2(config=tacotron_config, training=True, name="tacotron2")
        tacotron2._build()
        tacotron2.summary()
        # for finetune
        tacotron2.load_weights("examples/tacotron2/exp/A/checkpoints/model-200000.h5")`

and loss output:
2020-08-10 15:08:42,265 (base_trainer:750) INFO: (Steps: 500) Finished evaluation (25 steps per epoch).
2020-08-10 15:08:42,266 (base_trainer:757) INFO: (Steps: 500) eval_stop_token_loss = nan.
2020-08-10 15:08:42,267 (base_trainer:757) INFO: (Steps: 500) eval_mel_loss_before = nan.
2020-08-10 15:08:42,268 (base_trainer:757) INFO: (Steps: 500) eval_mel_loss_after = nan.
2020-08-10 15:08:42,269 (base_trainer:757) INFO: (Steps: 500) eval_guided_attention_loss = nan.

seems the problem is caused by loss=nan

**Comment by dathudeptrai** [comment]
@unparalleled-ysj so let training from scratch ? and report here

**Comment by orikama** [comment]
I also faced this problem, with fine-tuning tacotron with 2 of my small custom datasets, however my third dataset works completely fine for some reason.  You can spot this NaN problem a lot earlier if you set `log_interval_steps` to 100 or 50.
I also found out that restarting training can help when you see that loss turns to NaN, although you can waste hours by doing that (because you may need 10+ restarts), and of course it doesn't fix the cause of the problem.
I feel like there is a lot of randomness to this, when I saw this issue being opened, I started fine-tuning again one of my datasets that has this NaN problem, just to check if problem still exist, and it worked fine.

**Comment by orikama** [comment]
@dathudeptrai Training from scratch is impossible in my case, because my datasets is too small for that. I will try your fix with `tf.keras.layers.BatchNormalization` when I'm done with fine-tuning your ljspeech tacotron2 checkpoint on arpabet phonemes.

**Comment by dathudeptrai** [comment]
@orikama thanks for report. I'm working on new base_trainer that handle nan problem better. The simple solution is that relace all samples which has value = nan or its loss = nan by 0.0 :)).

**Comment by unparalleled-ysj** [comment]
@dathudeptrai Training from scratch is okay, there will be no nan problem
![new](https://user-images.githubusercontent.com/39238619/89763315-d65dd280-db24-11ea-86ce-144ae246aafc.png)
![new_alignment](https://user-images.githubusercontent.com/39238619/89763333-e1186780-db24-11ea-9f36-6b733f361aa6.png)
but I want to use small data for fine-tuning. If you train from scratch, the amount of data is not enough

**Comment by dathudeptrai** [comment]
> @dathudeptrai Training from scratch is okay, there will be no nan problem
> ![new](https://user-images.githubusercontent.com/39238619/89763315-d65dd280-db24-11ea-86ce-144ae246aafc.png)
> ![new_alignment](https://user-images.githubusercontent.com/39238619/89763333-e1186780-db24-11ea-9f36-6b733f361aa6.png)
> but I want to use small data for fine-tuning. If you train from scratch, the amount of data is not enough

Let try above suggestion and see if it fix this problem. BTW, let try 1 model from scratch and see if it's still good enough. You want to use tacotron-2 or just training it to extract duration for fastspeech2 ?.

**Comment by unparalleled-ysj** [comment]
@dathudeptrai ok, I will try it . I training tacotron2 just want extract duration for fastspeech2

**Comment by unparalleled-ysj** [comment]
@dathudeptrai BTW, when I was preprocessing, some utt caused this warning message: f"{utt_id} causes clipping. It is better to reconsider global gain scale value." Is this the key to causing nan and how should it be solved?

**Comment by dathudeptrai** [comment]
> @dathudeptrai BTW, when I was preprocessing, some utt caused this warning message: f"{utt_id} causes clipping. It is better to reconsider global gain scale value." Is this the key to causing nan and how should it be solved?

it's not a problem :D.

**Comment by orikama** [comment]
> BTW, i just feel the nan problem on tacotron cause by dataset-specific.

It can be, however, with [NVIDIA tacotron2](https://github.com/NVIDIA/tacotron2) i've never had such a problem.
> The simple solution is that relace all samples which has value = nan or its loss = nan by 0.0 :))

I don't know how to do that :) I'm a c++/c# programmer, I have a very basic knowledge of machine learning, tensorflow and python. I've tried to fix NaN loss like this (borrowed it from stackoverflow) :
```
if tf.math.is_nan(loss_att):
    def replacenan(t):
        return tf.where(tf.math.is_nan(t), tf.zeros_like(t), t)
    loss_att = replacenan(loss_att)
```
but it gives error when I start training.

**Comment by orikama** [comment]
>@dathudeptrai BTW, when I was preprocessing, some utt caused this warning message: f"{utt_id} causes clipping. It is better to reconsider global gain scale value." Is this the key to causing nan and how should it be solved?

yeah, I didn't have issues with preprocessing and still get NaN loss

**Comment by dathudeptrai** [comment]
> > BTW, i just feel the nan problem on tacotron cause by dataset-specific.
> 
> It can be, however, with [NVIDIA tacotron2](https://github.com/NVIDIA/tacotron2) i've never had such a problem.
> 
> > The simple solution is that relace all samples which has value = nan or its loss = nan by 0.0 :))
> 
> I don't know how to do that :) I'm a c++/c# programmer, I have a very basic knowledge of machine learning, tensorflow and python. I've tried to fix NaN loss like this (borrowed it from stackoverflow) :
> 
> ```
> if tf.math.is_nan(loss_att):
>     def replacenan(t):
>         return tf.where(tf.math.is_nan(t), tf.zeros_like(t), t)
>     loss_att = replacenan(loss_att)
> ```
> 
> but it gives error when I start training.

i will commit the new_trainer tonight or tmr to try to fix this nan problem :)). Not sure if it can because i can't reproduce nan problem in my experiment ^^. BTW, can you compared my tacotron-2 quality with nvidia tacotron2?. :D.

**Comment by orikama** [comment]
> can you compared my tacotron-2 quality with nvidia tacotron2?

It is hard to compare them, because I used nvidia tacotron2 + waveglow(not finetuned), and now I'm using your tacotron2 + [WaveRNN](https://github.com/fatchord/WaveRNN). But overall I think both your and nvidia implementation gives almost the same result.
Plus I've only trained nvidia implementation with only 1 of my 3 datasets a long time ago. With my new dataset (which is only 55 minutes), you repo gives a pretty good result.
Main difference is, nvidia repo trains a lot faster, especially if I turn on fp16 training (with which I can set batch_size = 60), although I feel that your model converges faster, but honestly I don't remember, it was half a year ago, may be training speed is actually the same. 
And I hope that when I retrain my models with phonemes it will be a little bit better :)

**Comment by unparalleled-ysj** [comment]
> @unparalleled-ysj so let training from scratch ? and report here. BTW the problem may because batch_norm (#187). Pls relace all `tf.keras.layers.experimental.SyncBatchNormalization` in https://github.com/TensorSpeech/TensorFlowTTS/blob/master/tensorflow_tts/models/tacotron2.py to `tf.keras.layers.BatchNormalization`

using 'tf.keras.layers.BatchNormalization' to replace 'tf.keras.layers.experimental.SyncBatchNormalization' still encount nan problem

**Comment by dathudeptrai** [comment]
> > @unparalleled-ysj so let training from scratch ? and report here. BTW the problem may because batch_norm (#187). Pls relace all `tf.keras.layers.experimental.SyncBatchNormalization` in https://github.com/TensorSpeech/TensorFlowTTS/blob/master/tensorflow_tts/models/tacotron2.py to `tf.keras.layers.BatchNormalization`
> 
> using 'tf.keras.layers.BatchNormalization' to replace 'tf.keras.layers.experimental.SyncBatchNormalization' still encount nan problem

try other checkpoint, Tacotron-(40k, 65k, 75k, 85k, 100k, 120k) [here](https://drive.google.com/drive/u/1/folders/1FQG8XC5c5JJ0jpCUl7Oqu5u2fC8IarP3)

**Comment by tekinek** [comment]
@unparalleled-ysj  As in #125, resuming from a step where nan was about to happen worked for me. I guess this NaN is somehow dataset related. Some samples together as a batch can trigger NaN loss (batch normalization problem?). Resuming may prevent such a batch comes again, and if you are luckly, no NaN happens. But this is also dataset related trick. I did not have such problem with Mozilla TTS implementations of tacotron2.

@dathudeptrai  I did a simple comparison on tacotron2 from this repo and Mozilla TTS on my private dataset. But since I still have a training problem with TesorflowTTS mb-melgan, I used GL to produce waveform. 

My impression is:
1) Skipping or repeating a word: Less likely happen in TensorflowTTS compared to near-vanilla implementation in Mozilla TTS (master), but on a par with its updated version which uses double decoder consistency (dev branch).

2) Stop early or at maximum decoder steps: Never happened with TensorflowTTS, but often with Mozilla TTS (master). Dev branch somehow solved it. Usage of eos in TensorflowTTS may helped a lot in this case, where Mozilla TTS uses it only for phoneme-based training, but not for character-based one. Now I'm also training Mozillat TTS by attaching eos manually, so far stopnet loss looks much better than before.

3) Naturalness: Mozillat TTS (dev) produces more natural speech than TensorflowTTS with same reduction_factor, and it still improves with more training (like 400k). TensorflowTTS seems over-fit more quickly.

This is not a serious comparison, but just an impression based on hearing couple of sentences. Hope you spend some time with tacotron2 in this repo. It is slow but synthesis quality is solid. Double decoder consistency seems a good addition:)

**Comment by dathudeptrai** [comment]
@tekinek thank for ur insight. Actually, as you saw, the name
Of the framework focus to real-time that is why i spent more tỉme for fastspeech and fastspeech2 rather than tacotron. The reason why i still implement tacotron2 is because at the time i implement this framework, tacotron2 is still need to extract duration for fastspeech. So i just need to make my tacotron2 learn aignment as fast as póible. I think my model convergence so fast rather than overfit too soon because base on the loss value, it is very good compared with other implementation. But the loss value is not linear correlation with quality, it is based on many factors. When i implement any models, i just try my best to make it better or atleast on par with existing implementation and in my thought, the quality of ljspeech is ok when compared with other implementation. If i have tỉme i ưill try to make tacotron2 better, now i am implementing parallelwavegan :))z

**Comment by tekinek** [comment]
@dathudeptrai  Looking forward to your parallelwavegan:)

**Comment by ZDisket** [comment]
I also have NaNs when finetuning Tacotron2, happens randomly in my Spanish single-speaker dataset I'm trying to train, but 100% of the when it's any of my multispeaker datasets.

**Comment by dathudeptrai** [comment]
@unparalleled-ysj is this problem still ur problem ?

**Comment by unparalleled-ysj** [comment]
@dathudeptrai I did not try tacotron2 fine-tune again, now I use MFA to extract durations
